fairlearn/metrics/_metrics_engine.py

```diff
@@ -82,33 +95,114 @@ def _check_metric_params(y_true, metric_params,
-                             y_true,
-                             y_pred,
+
+    def __init__(self, metric_function, indexed_params=None, name=None):
+        self._metric_function = metric_function
+        self._indexed_params = indexed_params
+        if name is not None:
+            self.__name__ = name
+
+    def __repr__(self):
+        if self._indexed_params is None:
+            args_string = ""
+        else:
+            args_string = ", indexed_params={0}".format(self._indexed_params)
+        return "make_metric_group_summary({0}{1})".format(
+            _function_name(self._metric_function),
+            args_string)
+
+    def __call__(self, y_true, y_pred, *, sensitive_features, **metric_params):
+        return group_summary(self._metric_function,
+                             y_true, y_pred,
                              sensitive_features=sensitive_features,
-                             indexed_params=indexed_params,
+                             indexed_params=self._indexed_params,
                              **metric_params)
 
-    # Improve the name of the returned function
-    wrapper.__name__ = "group_{0}".format(metric_function.__name__)
 
-    return wrapper
+class _DerivedMetricCallable:
+    """Callable that calculates a derived metric.
+
+    :param transformation_function: A transformation function with the signature
+        ``transformation_function(summary)``
+    :type transformation_function: func
+
+    :param summary_function: A metric group summary function with the signature
+        ``summary_function(y_true, y_pred, *, sensitive_features, **metric_params)``
+    :type summary_function: func
+    """
+
+    def __init__(self, transformation_function, summary_function, name=None):
+        self._transformation_function = transformation_function
+        self._summary_function = summary_function
+        if name is not None:
+            self.__name__ = name
+
+    def __repr__(self):
+        return "make_derived_metric({0}, {1})".format(
+                _function_name(self._transformation_function),
+                _function_name(self._summary_function))
+
+    def __call__(self, y_true, y_pred, *, sensitive_features, **metric_params):
+        return self._transformation_function(self._summary_function(
+                y_true, y_pred,
+                sensitive_features=sensitive_features,
+                **metric_params))
+
+
+def make_metric_group_summary(metric_function, indexed_params=None, name=None):
+    """Make a callable that calculates the group summary of a metric.
+
+    :param metric_function: A metric function with the signature
+        ``metric_function(y_true, y_pred, **metric_params)``
+    :type metric_function: func
+
+    :param indexed_params: The names of parameters of ``metric_function`` that
+        should be split according to ``sensitive_features`` in addition to ``y_true``
+        and ``y_pred``. Defaults to ``None`` corresponding to ``['sample_weight']``.
```

Comment:
I've just spotted this, and.... should this parameter be `extra_indexed_params`? What happens if the user specifies a list which doesn't include `sample_weight` ?

Reply:
My sense is that the user would only supply this after they inspect the signature of the base metric in question, and if `sample_weight` is not part of the signature, I think it would be weird to *always* include it... An alternative would be for the default value to be `['sample_weight']` rather than `None`. But given that the behavior of that is equivalent, I would leave that to a future PR.

Reply:
I'm not a huge fan, but I'm not sure which would be considered more Pythonic.